openTSNE/nearest_neighbors.py

```diff
@@ -369,22 +369,55 @@ def build(self, data, k):
         # unless we're actually going to use it
         import pynndescent
 
+        # Will use query() only for k>15
+        if k <= 15:
+            n_neighbors_build = k + 1
+        else:
+            n_neighbors_build = 15
+
+        # due to a bug, pynndescent currently does not support n_jobs>1
+        # for sparse inputs. This should be removed once it's fixed.
+        n_jobs_pynndescent = self.n_jobs
+        import scipy.sparse as sp
+
+        if sp.issparse(data):
+            n_jobs_pynndescent = 1
+
         # UMAP uses the "alternative" algorithm, but that sometimes causes
         # memory corruption, so use the standard one, which seems to work fine
         self.index = pynndescent.NNDescent(
             data,
-            n_neighbors=15,
+            n_neighbors=n_neighbors_build,
             metric=self.metric,
             metric_kwds=self.metric_params,
             random_state=self.random_state,
             n_trees=n_trees,
             n_iters=n_iters,
             algorithm="standard",
             max_candidates=60,
-            n_jobs=self.n_jobs,
+            n_jobs=n_jobs_pynndescent,
         )
 
-        indices, distances = self.index.query(data, k=k + 1)
+        # -1 in indices means that pynndescent failed
+        indices, distances = self.index.neighbor_graph
+        mask = np.sum(indices == -1, axis=1) > 0
+
+        if k > 15:
+            indices, distances = self.index.query(data, k=k + 1)
+
+        # as a workaround, we let the failed points group together
+        if np.sum(mask) > 0:
+            warnings.warn(
+                f"`pynndescent` failed to find neighbors for some of the points."
+                f"As a workaround, openTSNE considers all such points similar to "
+                f"each other, so they will likely form a cluster in the embedding."
+            )
+            distances[mask] = 1
+            fake_indices = np.random.choice(
+                np.sum(mask), size=np.sum(mask) * indices.shape[1]
```

Comment:
I checked that the default for `np.random.choice` is `replace=True`, so this will work even if there aren't enough failed points. However, I still had to check since I don't know this by heart. Setting `replace=True` explicitly would be clearer and make it future proof, in case numpy ever decides to change their defaults.